ocs_ci/deployment/acm.py

```diff
@@ -0,0 +1,192 @@
+import shlex
+import pexpect
+
+from ocs_ci.framework import config
+from ocs_ci.ocs import constants
+from ocs_ci.ocs.constants import SUBMARINER_GATEWAY_PROMPT
+from ocs_ci.ocs.exceptions import (
+    CommandFailed,
+    DRPrimaryNotFoundException,
+    InteractivePromptException,
+)
+from ocs_ci.utility.utils import run_cmd
+
+
+logger = logging.getLogger(__name__)
+
+
+def run_subctl_cmd(cmd=None):
+    """
+    Run subctl command
+
+    Args:
+        cmd: subctl command to be executed
+
+    Returns:
+
+    """
+    cmd = " ".join("subctl", cmd)
+    out = run_cmd(cmd)
+    logger.info(out)
+
+
+def run_subctl_cmd_interactive(cmd, prompt, answer):
+    """
+    Handle interactive prompts with answers during subctl command
+
+    Args:
+        cmd(str): Command to be executed
+        prompt(str): Expected question during command run which needs to be provided
+        answer(str): Answer for the prompt
+
+    Raises:
+        InteractivePromptException: in case something goes wrong
+
+    """
+    cmd = " ".join("subctl", cmd)
+
+    child = pexpect.spawn(cmd)
+    if child.expect(prompt, timeout=config.ENV_DATA["submariner_prompt_timeout"]):
+        raise InteractivePromptException("Unexpected Prompt")
+
+    if not child.sendline("".join(answer, constants.ENTER_KEY)):
+        raise InteractivePromptException("Failed to provide answer to the prompt")
+
+
+class Submariner(object):
+    """
+    Submariner configuaration and deployment
+    """
+
+    def __init__(self, index):
+        # whether upstream OR downstream
+        self.source = config.ENV_DATA["submariner_source"]
+        # Designated broker cluster index where broker will be deployed
+        self.designated_broker_cluster_index = self.get_primary_cluster_index()
+        # sequence number for the clusters from submariner perspective
+        # Used mainly to run submariner commands, for each cluster(except ACM hub) we will
+        # assign a seq number with 1 as primary and continue with subsequent numbers
+        self.cluster_seq = 1
+        # List of index to all the clusters which are participating in DR (except ACM)
+        # i.e index in the config.clusters list
+        self.dr_only_list = []
+
+    def deploy(self):
+        if self.source == "upstream":
+            self.deploy_upstream()
+        else:
+            self.deploy_downstream()
+
+    def deploy_upstream(self):
+        self.download_binary()
+        self.submariner_configure_upstream()
+
+    def deploy_downstream(self):
+        pass
+
+    def download_binary(self):
+        if self.source == "upstream":
+            # This script puts the platform specific binary in ~/.local/bin
+            # we need to move the subctl binary to ocs-ci/bin dir
+            download_cmd = "curl -Ls https://get.submariner.io | bash"
+            try:
+                run_cmd(download_cmd)
+            except CommandFailed:
+                logger.exception("Failed to download submariner binary")
```

Comment:
done, handled both download and run failures.

Reply:
@Avilir , there are 2 parts to this function
1. Curl failures: may be because of network down or something
2. curl will actually fetch a downloader script which will then download submariner binary

So I hope both exceptions convey it. 
Note: We are not running submariner command here but just the download